compiler/z/codegen/OMRCodeGenerator.cpp

```diff
@@ -2475,11 +2476,96 @@ OMR::Z::CodeGenerator::generateScratchRegisterManager(int32_t capacity)
+   // label instruction will be prepended to the start of the instruction stream.
+   _methodBegin = new (self()->trHeapMemory()) TR::S390LabelInstruction(TR::InstOpCode::LABEL, self()->comp()->getStartTree()->getNode(), generateLabelSymbol(self()), static_cast<TR::Instruction*>(NULL), self());
+   
+   _methodEnd = generateS390LabelInstruction(self(), TR::InstOpCode::LABEL, self()->comp()->findLastTree()->getNode(), generateLabelSymbol(self()));
+
    TR_S390BinaryEncodingData data;
+   data.cursorInstruction = self()->getFirstInstruction();
+   data.estimate = 0;
    data.loadArgSize = 0;
-   self()->fe()->generateBinaryEncodingPrologue(&data, self());
+   TR::Recompilation* recomp = self()->comp()->getRecompilationInfo();
+
+   //  setup cursor for JIT to JIT transfer
+   //
+   if (self()->comp()->getJittedMethodSymbol()->isJNI() && !self()->comp()->getOption(TR_FullSpeedDebug))
+      {
+      data.preProcInstruction = TR::Compiler->target.is64Bit() ?
+         data.cursorInstruction->getNext()->getNext()->getNext() : 
+         data.cursorInstruction->getNext()->getNext();
+      }
+   else
+      {
+      data.preProcInstruction = data.cursorInstruction->getNext();
+      }
+
+   data.jitTojitStart = data.preProcInstruction->getNext();
+
+   // Generate code to setup argument registers for interpreter to JIT transfer
+   // This piece of code is right before JIT-JIT entry point
+   //
+   TR::Instruction * preLoadArgs, * endLoadArgs;
+   preLoadArgs = data.preProcInstruction;
+
+   // We need full prolog if there is a call or a non-constant snippet
+   //
+   TR_BitVector * callBlockBV = self()->getBlocksWithCalls();
+
+   // No exit points, hence we can
+   //
+   if (callBlockBV->isEmpty() && !self()->anyNonConstantSnippets())
+      {
+      self()->setExitPointsInMethod(false);
+      }
+
+   endLoadArgs = self()->getLinkage()->loadUpArguments(preLoadArgs);
+
+   if (recomp != NULL)
+      {
+#ifdef J9_PROJECT_SPECIFIC
+      if (preLoadArgs != endLoadArgs)
+         {
+         data.loadArgSize = CalcCodeSize(preLoadArgs->getNext(), endLoadArgs);
+         }
+
+      ((TR_S390Recompilation *) recomp)->setLoadArgSize(data.loadArgSize);
+#endif
+      recomp->generatePrePrologue();
+      }
+#ifdef J9_PROJECT_SPECIFIC
+   else if (self()->comp()->getOption(TR_FullSpeedDebug) || self()->comp()->getOption(TR_SupportSwitchToInterpreter))
+      {
+      self()->generateVMCallHelperPrePrologue(self()->getFirstInstruction());
+      }
+#endif
+
+   data.cursorInstruction = self()->getFirstInstruction();
+
+   // Padding for JIT Entry Point
+   //
+   if (!self()->comp()->compileRelocatableCode())
+      {
+      data.estimate += 256;
+      }
+
+   TR::Instruction* cursor = data.cursorInstruction;
+
+   // TODO: We should be caching the PROC instruction as it's used in several places and is pretty important
+   while (cursor && cursor->getOpCodeValue() != TR::InstOpCode::PROC)
```

Comment:
I also want to consolidate this so it's the same on all platforms. I'm working on making such a change at the common level.